### PR TITLE
Add support for Collada references

### PR DESCRIPTION
**Pull request by lukicdarkoo**
Tasks:
- [x] Add ColladaShapes PROTO
- [x] Add a name flag to the Mesh node (`name` corresponds the `name` attribute in Collada files) 
  - [x] Test performances, Mesh vs IndexFaceSet (he version with Mesh nodes was way ahead of the IndexFacedSet version, so I didn't bother to quantify it)
- [x] Add documentation
- [x] Avoid circular dependencies
- [x] Add changelog
- [x] Fix crashes when choosing the file
- [x] Check if the sub-mesh exists

Docs:
- https://www.cyberbotics.com/doc/guide/object-shapes?version=feature-collada-mesh

### CODE REVIEW

**Comment by omichel** [comment]
If too difficult to implement in Webots, we could also implement it as a PROTO that would instanciate a Group of Shape nodes with PBRAppearence and IndexedFaceSet.

**Comment by lukicdarkoo** [comment]
> If too difficult to implement in Webots, we could also implement it as a PROTO that would instanciate a Group of Shape nodes with PBRAppearence and IndexedFaceSet.

It would certainly take a lot of time to implement such node in Webots. Creating a PROTO sounds like a good idea, but I am not sure what is a good way to generate PBRAppearence. Instead of IndexedFaceSet we can use Mesh and just pass .dae path to it

**Comment by omichel** [comment]
> Instead of IndexedFaceSet we can use Mesh and just pass .dae path to it

This won't work as the Mesh node will merge all the submeshes together which is what we want to avoid. Or we should add an extra parameter to the Mesh node that indicates a submesh.

**Comment by omichel** [comment]
To further clarify my idea, there are two things:
1) Add an "id" SFString parameter to the Mesh node that indicates a specific mesh inside a collada file (the "id" field will be ignored in case of OBJ or STL file).
2) Create a new ColladaShapes proto that will construct the collada shapes, including appearances and meshes, generating the following structure:
```js
Group {
  children [
    Shape {
      appearance PBRAppearance {
        # green appearance
      }
      geometry Mesh {
        url "collada.dae"
        id "1"
      }
    }
    Shape {
      appearance PBRAppearance {
        # red color
      }
      geometry Mesh {
        url "collada.dae"
        id "2"
      }
    }
    ...
  ]
}
```
Basically, the proto file will use Javascript code to open and load the content of the "collada.dae" file. It will extract the appearance properties and create the corresponding PBRAppearance node which will be associated with specific meshes inside the collada file.

Another option would be to create IndexedFaceSet nodes instead of Mesh nodes, but I fear this option would have a strong performance impact at loading time...

**Comment by lukicdarkoo** [comment]
@omichel We have the prototype ready. I believe the solution is slightly better than the one we talked about, there is no code duplication, heavy lifting is done on the C++ side, and there are no many changes.

The Mesh optimization needs to be implemented (and current performance evaluated).

**Comment by omichel** [comment]
That is a very nice idea to implement it this way.

**Comment by lukicdarkoo** [comment]
@omichel I don't see a way to avoid circular dependency, please let me know if you have any idea.

**Comment by omichel** [comment]
Can't we move `WbQjsCollada.cpp` into `nodes/utils`?

**Comment by lukicdarkoo** [comment]
> Can't we move `WbQjsCollada.cpp` into `nodes/utils`?

I worked around it with a forward declaration. We typically forward declare only the class, but here I included a method as well.

File: src/webots/nodes/WbMesh.cpp
Comment (ad-daniel) [inline src/webots/nodes/WbMesh.cpp:36]: A connection like `connect(mName, &WbSFString::changed, this, &WbMesh::updateName);` should be added. We should probably also check that the name exists within the collada file.

File: docs/reference/changelog-r2021.md
Comment (ad-daniel) [inline docs/reference/changelog-r2021.md:14]: This branch needs to be resynched with develop and the changelog entry should be added to `changelog-r2022.md` instead.

File: docs/reference/mesh.md
Comment (ad-daniel) [inline docs/reference/mesh.md:34]: From the sub-meshes (plural) it leads to believe it's possible to specify more than one, but the parameter is of type `SFString`. Is it supposed to be possible to import say 2 meshes at once in the same ColladaShape or the actual options are: all of the meshes OR just one.

**Review by ad-daniel — CHANGES_REQUESTED** [review]
Didn't test it yet, but got a few comments

**Review by omichel — CHANGES_REQUESTED** [review]
I am not yet convinced that re-declaring the class is a good solution as I fear that this may cause memory leaks (because the actual object is not the one we believe it is, for example, I am not sure sizeof(WbQjsCollada) is the same when called from WbTemplateEngine.cpp and from WbQjsCollada.cpp). Also it introduces some code redundancy...

File: projects/shapes/protos/ColladaShapes.proto
Comment (lukicdarkoo) [reply projects/shapes/protos/ColladaShapes.proto:25]: I used a style from the [WoodenBox](https://github.com/cyberbotics/webots/blob/master/projects/objects/factory/containers/protos/WoodenBox.proto) PROTO. Maybe we could reformat other PROTOs as well.

File: docs/reference/mesh.md
Comment (lukicdarkoo) [reply docs/reference/mesh.md:34]: It will include all sub-meshes with the given name (I am not sure if the same names are allowed). However, I was thinking to support the * (asterisk) wildcard, I am not sure if anyone will need it.

**Comment by lukicdarkoo** [comment]
> I am not yet convinced that re-declaring the class is a good solution as I fear that this may cause memory leaks (because the actual object is not the one we believe it is, for example, I am not sure sizeof(WbQjsCollada) is the same when called from WbTemplateEngine.cpp and from WbQjsCollada.cpp). Also it introduces some code redundancy...

I don't keen on using it either as I am not aware of the side effects. As an alternative, we can put the header file in `core` and keep the implementation in `node/utils`. However, it will be also the first time we are separating .hpp/.cpp files.

File: docs/guide/object-shapes.md
Comment (lukicdarkoo) [reply docs/guide/object-shapes.md:3]: I don't think it is needed as this is a PROTO, not a node

File: docs/guide/object-shapes.md
Comment (lukicdarkoo) [reply docs/guide/object-shapes.md:8]: Maybe, it is more accurate `into a single one` -> `into a single shape`?

File: src/webots/nodes/WbMesh.cpp
Comment (lukicdarkoo) [reply src/webots/nodes/WbMesh.cpp:35]: Good point, implemented

File: docs/guide/object-shapes.md
Comment (omichel) [reply docs/guide/object-shapes.md:8]: It's not a single shape, it's a Group of Shape nodes, so I would say:
"As Collada files may contain several meshes with material information, each mesh will be added as a Shape node inside a Group node."

File: docs/reference/mesh.md
Comment (omichel) [reply docs/reference/mesh.md]: I believe that if the `name` string is empty (default value), it should include all the meshes instead of none to be more user friendly. Therefore, the wildcard "*" would give the same behavior as the empty string, so maybe we can drop it for now and document only the behavior with the empty string. Otherwise, people will expect that strings like "*_arm" to work, e.g., match "left_arm" and "right_arm". In such a case I am not sure which material to use... The one of the right or left arm? So, I would postpone the implementation of the wildcards until we really need them.

Comment (omichel) [reply docs/reference/mesh.md]: I am not sure how Collada files are organized and if several meshes can use the same appearance, but it is the case, it would be more logical to merge all the meshes using the same appearance into the same IndexedFaceSet inside a Shape node. In that case the "name" parameter should refer to the appearance and not to the mesh... But I am not sure if appearance are named in Collada files...

File: projects/shapes/protos/ColladaShapes.proto
Comment (omichel) [reply projects/shapes/protos/ColladaShapes.proto]: Yes, but we should do it in a separate PR.

**Comment by omichel** [comment]
Can't we simply have some static method in WbQjsCollada to avoid creating a new singleton class like WbQjsColladaRelay?

**Comment by lukicdarkoo** [comment]
> Can't we simply have some static method in WbQjsCollada to avoid creating a new singleton class like WbQjsColladaRelay?

Yes, perfect, thanks!

**Review by omichel — APPROVED** [review]
It looks very good now, thank you.

**Review by omichel — APPROVED** [review]
Great, thank you.

**Comment by omichel** [comment]
@ad-daniel: we need your approval before we can merge this PR.

**Review by ad-daniel — APPROVED** [review]
Thank you.

**Pull request merged**
